File: Problems/42_trappingRainWater.ts
function trap(height: number[]): number {
    //front and back pointer, will attempt to fill optimally

    let [front, back] = [0, height.length - 1];

    let [frontMaxHeight, backMaxHeight] = [0, 0]; //could also be height[front], height[back]

    let totalArea = 0;

    while (front < back) {
        const [frontHeight, backHeight] = [height[front], height[back]];

        //now if there was nothing between front and back, we could fill to a max height of min(front, back)
        //therefore to fill optimatlly we must look at the limiting height
        if (frontHeight < backHeight) {
            //front height is limiting
            //now the max height we can fill from this point is upto the maximum prior left height
            if (frontHeight < frontMaxHeight) { //there was a 'dip'
                totalArea += frontMaxHeight - frontHeight;
            }
            else {
                //otherwise there was no dip, and therefore no area can be filled
                frontMaxHeight = frontHeight
            }
            front += 1;
        }
        else { // back height is limiting
            //similar idea to front height, and by only moving the limiting height, we ensure we won't reach a situation where the front pointer is filling to a greater height than the back pointer or vice versa

            if (backHeight < backMaxHeight) {
                totalArea += backMaxHeight - backHeight;
            }
            else {
                backMaxHeight = backHeight;
            }
            back -= 1;
        }
    }

    return totalArea;
};

console.log(trap([0,1,0,2,1,0,1,3,2,1,2,1]));

File: Problems/42_trappingRainWater.py
class Solution(object):
    def trap(self, height):
        #use 2 pointers
        #first pointer is base pointer, and keeps track of current elevation level
        #second pointer is look ahead pointer, and continues ahead until it reaches a height equal to or greater than the base pointer
        #water will be stored after the base pointer if the next height is lower than it (as long as we do eventually reach another base pointer to act as the end marker of the interval)

        def baseLookAhead(subheight):
            totalArea = 0

            base = 0
            while base < len(subheight):
                currentBatchArea = 0
                lookAhead = base + 1
                while lookAhead < len(subheight) and subheight[lookAhead] < subheight[base]:
                    currentBatchArea += subheight[base] - subheight[lookAhead]
                    lookAhead += 1

                if lookAhead != len(subheight):
                    totalArea += currentBatchArea #if lookAhead is equal to the length of heights, it means there was no end-interval, so we cannot add the final batch area

                base = lookAhead

            return totalArea

        #Use base lookAhead method, except we will split the input heights into 2 sublists based around the maximum height (since this can always be used as an end-interval
        maxIndex = height.index(max(height))
        firstHalf = height[0 : maxIndex + 1]
        secondHalf = height[maxIndex : len(height)][::-1] #reverse to ensure the max height is always at the end interval

        return baseLookAhead(firstHalf) + baseLookAhead(secondHalf)

solution = Solution()
print(solution.trap([0,1,0,2,1,0,1,3,2,1,2,1]))